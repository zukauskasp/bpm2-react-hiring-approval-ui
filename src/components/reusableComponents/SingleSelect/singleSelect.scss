.formWrapper {

    .MuiSvgIcon-root {
        font-size: 2.4rem;
    }

    .MuiNativeSelect-icon {
        right: 8px;
        top: calc(50% - 16px);
        color: #003755;
    }
}