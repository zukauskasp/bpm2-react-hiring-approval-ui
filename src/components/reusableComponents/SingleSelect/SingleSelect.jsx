import React, { useState } from "react";
import "./singleSelect.scss";
import { makeStyles, withStyles } from '@material-ui/core/styles';
import PropTypes from 'prop-types';

import Tooltip from "../Tooltip/Tooltip";
import FormControl from '@material-ui/core/FormControl';
import NativeSelect from '@material-ui/core/NativeSelect';
import InputBase from '@material-ui/core/InputBase';
import KeyboardArrowDownIcon from '@material-ui/icons/KeyboardArrowDown';
import KeyboardArrowUpIcon from '@material-ui/icons/KeyboardArrowUp';

const SingleSelect = ({ label, required, value, onChange, items, errorMessage, ...otherProps }) => {

    const classes = useStyles();
    const [isOpen, setIsOpen] = useState(false);
    const handleSelectArrows = () => setIsOpen(!isOpen);

    return (
        <FormControl className={`${classes.formWrapper} formWrapper`}>
            {label ? <label className={classes.label}>{label} {required ? " *" : null}</label> : null}
            <NativeSelect
                {...otherProps}
                value={value}
                onChange={onChange}
                onClick={handleSelectArrows}
                input={<BootstrapInput />}
                IconComponent={isOpen ? KeyboardArrowUpIcon : KeyboardArrowDownIcon}>
                {items ? items.map((option, index) => (
                    <option value={option.value} key={index}> {option.name}</option>
                )) : null}
            </NativeSelect>
            {errorMessage ? <Tooltip text={errorMessage} messageType='error' /> : null}
        </FormControl>
    );
}


SingleSelect.propTypes = {
    errorMessage: PropTypes.string,
    required: PropTypes.bool,
    label: PropTypes.string,
    onChange: PropTypes.func.isRequired,
    items: PropTypes.arrayOf(PropTypes.exact({
        name: PropTypes.string,
        value: PropTypes.string
    })).isRequired,
    value: PropTypes.string,
}

SingleSelect.defaultProps = {
    errorMessage: "",
    label: ""
}


export default SingleSelect;

const BootstrapInput = withStyles(theme => ({
    root: {
        width: '280px',
    },
    input: {
        borderRadius: 4,
        color: '#495057',
        position: 'relative',
        backgroundColor: theme.palette.background.paper,
        border: '1px solid #ced4da',
        fontSize: '16px',
        height: '48px',
        lineHeight: "48px",
        padding: '0 .75rem',
        transition: theme.transitions.create(['border-color', 'box-shadow']),
        boxSizing: "border-box",
        fontFamily: "Danske Text v2",
        '&:focus': {
            borderColor: 'rgb(0, 159, 218)',
            outline: 'none',
            borderRadius: 4
        },
    },
}))(InputBase);

const useStyles = makeStyles({
    formWrapper: {
        marginBottom: "24px"
    },
    label: {
        marginBottom: '8px',
        fontFamily: 'Danske Text v2',
        color: "#003755",
    }
});
