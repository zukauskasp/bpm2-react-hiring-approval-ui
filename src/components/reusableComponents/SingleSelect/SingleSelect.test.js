import React from 'react';
import { createShallow, createMount } from '@material-ui/core/test-utils';

import SingleSelect from './SingleSelect';

describe('SingleSelect', () => {

    let wrapper;
    let mount;
    let shallow;

    beforeAll(() => {
        shallow = createShallow();
        mount = createMount();
    });

    const list = [
        { name: "one", value: "1" },
        { name: "two", value: "2" },
        { name: "three", value: "3" },
    ]
    const props = {
        items: list,
        required: false,
        label: "Select number",
        onChange: () => { },
        errorMessage: "Error message"
    }

    beforeEach(() => {
        wrapper = shallow(<SingleSelect {...props} />);
    })

    afterAll(() => {
        mount.cleanUp();
    });

    it('expect to match latest snapshot', () => {
        expect(shallow(<SingleSelect {...props} />).debug()).toMatchSnapshot();
    });

    it('expect to render component', (() => {
        expect(wrapper.length).toEqual(1);
    }));

    it('expect to render correct props', () => {
        expect(mount(<SingleSelect {...props} />).props()).toEqual(props);
    });

    it('expect to render default props', () => {
        const mounted = mount(<SingleSelect items={props.items} onChange={props.onChange} />);

        expect(mounted.props().errorMessage).toEqual("");
        expect(mounted.props().label).toEqual("");
    });

    it('expect to render list', () => {
        const mounted = mount(<SingleSelect {...props} />);

        expect(mounted.prop('items').length).toEqual(list.length);
    });


})