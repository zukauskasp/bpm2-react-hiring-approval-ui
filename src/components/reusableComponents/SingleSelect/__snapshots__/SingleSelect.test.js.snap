// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleSelect expect to match latest snapshot 1`] = `
"<WithStyles(ForwardRef(FormControl)) className=\\"makeStyles-formWrapper-1 formWrapper\\">
  <label className=\\"makeStyles-label-2\\">
    Select number
     
  </label>
  <WithStyles(ForwardRef(NativeSelect)) value={[undefined]} onChange={[Function: onChange]} onClick={[Function: handleSelectArrows]} input={{...}} IconComponent={{...}}>
    <option value=\\"1\\">
       
      one
    </option>
    <option value=\\"2\\">
       
      two
    </option>
    <option value=\\"3\\">
       
      three
    </option>
  </WithStyles(ForwardRef(NativeSelect))>
  <Message text=\\"Error message\\" messageType=\\"error\\" position=\\"top-left\\" />
</WithStyles(ForwardRef(FormControl))>"
`;
